fix pprint of polyvariants that start with a core_type,closed, not low

chetmurthy opened the pull request:
print polyvariants that start with a core_type,closed, not low with leading bar ("|").

a type "[ | w ]" must be printed with the "|", or it won't be
reparseable.

with tests.

Octachron commented on parsing/parsetree.mli:
```diff
@@ -174,7 +174,7 @@ and row_field_desc =
         *)
   | Rinherit of core_type
-        (* [ T ] *)
+        (* [ | T ] *)
```
Since `T` is not a valid identifier for a type, maybe it would be better to update the documentation to `[ | t ] `.

Octachron commented on parsing/pprintast.ml:
```diff
@@ -326,6 +326,9 @@ and core_type1 ctxt f x =
           l longident_loc li
     | Ptyp_variant (l, closed, low) ->
+        let first_is_inherit = match l with
+            {Parsetree.prf_desc = Rinherit _}::_ -> true
```
Most of the time, relatively recent code in the compiler uses `|` even on the first clause of the pattern matching.

Octachron commented on testsuite/tests/printing-types/pprint_polyvariant.ml:
```diff
@@ -0,0 +1,17 @@
+{| type t = [ | w ] |} |> pa |> pr |> print_string ;; print_newline () ;;
+{| type t = [ | `A of int ] |} |> pa |> pr |> print_string ;; print_newline () ;;
+{| type t = [ `A of int ] |} |> pa |> pr |> print_string ;; print_newline () ;;
+
```
There is in fact a parsing round-trip test in `parsing/source.ml`. I think it would be better to move your test here.

chetmurthy commented:
Think I got 'em all.  Thank you for your careful review!  Ensuring all code meets the same standard is part of how caml-light was so successful.  *grin*

Octachron commented on parsing/pprintast.ml:
```diff
@@ -326,6 +326,9 @@ and core_type1 ctxt f x =
           l longident_loc li
     | Ptyp_variant (l, closed, low) ->
+        let first_is_inherit = match l with
+          | {Parsetree.prf_desc = Rinherit _}::_ -> true
```
The problematic case is narrower than that since:
```ocaml
type t = [ r | `B ]
```
parses just fine.
But printing the leading `|` whenever the first field is inherited seems fine.

Octachron commented on testsuite/tests/parsetree/source.ml:
```diff
@@ -7372,3 +7372,7 @@ let f = function
 
 let () =
   f (fun (type t) -> x)
+
```
I think it would be nice to add a comment with the issue and PR number to help future readers.

Octachron reviewed (approved):
The change is correct, as a small bug fix in the AST pretty-printer it seems fine for 4.11 .
@chetmurthy , would you mind adding a Change entry in the 4.11 section?

chetmurthy commented:
<strike>I, uh, think I understood the right place to put my Changes file update: -above- the 4.11 marker, right?  If not, let me know and I'll move it below.</strike> Oops, no, I got that wrong.  Fixing.

Octachron commented:
I visibly forgot about this PR during the 4.11 window. @chetmurthy , would you have the time to rebase this PR? I will then merge immediately.

chetmurthy commented:
Will fix and update branch per your suggestion.

chetmurthy committed:
re-do of print polyvariants that start with a core_type,closed, not low with leading bar ("|").

a type "[ | w ]" must be printed with the "|", or it won't be
reparseable.

with tests, Changes entry.

# Please enter the commit message for your changes. Lines starting
# with '#' will be ignored, and an empty message aborts the commit.
#
# On branch pr-polyvariant-pprint
# Changes to be committed:
#	modified:   Changes
#	modified:   parsing/parsetree.mli
#	modified:   parsing/pprintast.ml
#	modified:   testsuite/tests/parsetree/source.ml
#
# Untracked files:
#	Changes.orig
#	parsing/pprintast.ml.orig
#	testsuite/tests/parsetree/source.ml.orig
#	testsuite/tests/parsetree/source.ml.rej
#

chetmurthy commented:
I hope I did this correctly, and apologize if I screwed it up.  I tried to rebase twice, and got problems both times that I couldn't figure out (never rebase, so don't really have experience) and so I just took a diff and re-created the branch.

Octachron commented:
The recreated commit looks good. I will merge tomorrow morning. In this case, recreating the branch works fine too.

chetmurthy committed:
sorry, too-long line in Changes file

# Please enter the commit message for your changes. Lines starting
# with '#' will be ignored, and an empty message aborts the commit.
#
# On branch pr-polyvariant-pprint
# Your branch is up to date with 'my-fork/pr-polyvariant-pprint'.
#
# Changes to be committed:
#	modified:   Changes
#
# Untracked files:
#	Changes.orig
#	parsing/pprintast.ml.orig
#	testsuite/tests/parsetree/source.ml.orig
#	testsuite/tests/parsetree/source.ml.rej
#
